File: leetcode-solutions/ruby/307. Range Sum Query - Mutable.rb
# https://leetcode.com/problems/range-sum-query-mutable/

class NumArray

  =begin
      :type nums: Integer[]
  =end
      def initialize(nums)
          @nums = nums
          @bit = Array.new(nums.size + 1){0}
          prepare_bit
      end
  
  
  =begin
      :type i: Integer
      :type val: Integer
      :rtype: Void
  =end
      def add(i, val)
          current = i + 1
          
          while current < @bit.size
              @bit[current] += val
              current += (current & -current)
          end
      end
  
      def get_sum(i)
          current = i+1
          sum = 0
          while current >= 1
              sum += @bit[current]
              current -= (current & -current)
          end
          
          sum
      end
  =begin
      :type i: Integer
      :type j: Integer
      :rtype: Integer
  =end
      def sum_range(i, j)
          get_sum(j) - get_sum(i-1)
      end
  
      def prepare_bit
          @nums.each_with_index do |n, i|
              add(i, n)
          end
      end
      
      def update(i, n)
          diff = n - @nums[i]
          @nums[i] = n
          add(i, diff)
      end
  end
  
  # Your NumArray object will be instantiated and called as such:
  # obj = NumArray.new(nums)
  # obj.update(i, val)
  # param_2 = obj.sum_range(i, j)

File: leetcode-solutions/python/307. Range Sum Query - Mutable.py
# https://leetcode.com/problems/range-sum-query-mutable/
# use fenwick/binary indexed tree to solve this problem
# bit is for good for calculate sum query for one and two dimention array
# this is one based implementation
# n & -n extract last set bit to remaining part so 10: 1010 10 & -10 is 2(10)
# here adjustment is reset the lower set bit that is equvalent to
# i = i - (i & -i) i minux lower set bit that is reset lower set bit
class NumArray:
    def __init__(self, nums: List[int]):
        self.nums = nums
        self.size = len(nums)
        # one base int that's way size+1
        self.bit = [0 for _ in range(self.size+1)]
        for i in range(1, self.size+1):
          self._add(i, nums[i-1])

    def update(self, index: int, val: int) -> None:
      diff = val - self.nums[index]
      self.nums[index] = val
      # one base index: index+1
      self._add(index+1, diff)
        

    def sumRange(self, left: int, right: int) -> int:
      # since we want to include left that's why calling left
      return self._sum(right+1) - self._sum(left)
        
    def _sum(self, i):
      res = 0
      while i > 0:
        res += self.bit[i]
        i -= self._lowbit(i)
      return res
    
    def _lowbit(self, x):
      return x & -x
    
    def _add(self, i, val):
      while i <= self.size:
        self.bit[i] += val
        i += self._lowbit(i)

